examples/cryptocurrency/src/lib.rs

```diff
@@ -229,93 +215,97 @@ impl Transaction for TxTransfer {
-        let snapshot = self.blockchain.snapshot();
-        let schema = CurrencySchema::new(snapshot);
-        let idx = schema.wallets();
-        let wallets: Vec<Wallet> = idx.values().collect();
-
-        self.ok_response(&serde_json::to_value(&wallets).unwrap())
+impl Endpoint for GetWallet {
+    fn handle(&self, pubkey: PublicKey) -> Result<Wallet, ApiError> {
+        let schema = CurrencySchema::new(self.as_ref().snapshot());
+        schema.wallet(&pubkey).ok_or(ApiError::NotFound)
     }
+}
 
-    /// Common processing for transaction-accepting endpoints.
-    fn post_transaction(&self, req: &mut Request) -> IronResult<Response> {
-        match req.get::<bodyparser::Struct<CurrencyTransactions>>() {
-            Ok(Some(transaction)) => {
-                let transaction: Box<Transaction> = transaction.into();
-                let tx_hash = transaction.hash();
-                self.channel.send(transaction).map_err(ApiError::from)?;
-                let json = TransactionResponse { tx_hash };
-                self.ok_response(&serde_json::to_value(&json).unwrap())
-            }
-            Ok(None) => Err(ApiError::BadRequest("Empty request body".into()))?,
-            Err(e) => Err(ApiError::BadRequest(e.to_string()))?,
-        }
-    }
+read_request! {
+    /// Endpoint dumping all wallets from the storage.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # extern crate exonum;
+    /// # extern crate exonum_cryptocurrency;
+    /// #[macro_use] extern crate exonum_testkit;
+    /// #[macro_use] extern crate serde_json;
+    /// # use exonum::crypto;
+    /// # use exonum_cryptocurrency::{CurrencyService, TxCreateWallet, GetWallets};
+    /// use exonum_testkit::TestKit;
+    ///
+    /// # fn main() {
+    /// let mut testkit = TestKit::for_service(CurrencyService);
+    // We need to ensure that Alice's pubkey is lexicographically lesser than Bob's
+    // in order to assert the output of the endpoint.
+    /// # let mut keys: Vec<_> = (0..2).into_iter().map(|_| crypto::gen_keypair()).collect();
+    /// # keys.sort_by(|&(a, _), &(b, _)| a.cmp(&b));
+    /// # let (alice_pubkey, alice_key) = keys.swap_remove(0);
+    /// # let (bob_pubkey, bob_key) = keys.swap_remove(0);
+    /// # assert!(alice_pubkey <= bob_pubkey);
+    /// let tx_alice = TxCreateWallet::new(&alice_pubkey, "Alice", &alice_key);
+    /// let tx_bob = TxCreateWallet::new(&bob_pubkey, "Bob", &bob_key);
+    /// testkit.create_block_with_transactions(txvec![tx_alice, tx_bob]);
+    ///
+    /// testkit.api().test::<GetWallets>(
+    ///     json!(null),
+    ///     &json!([{
+    ///         "pub_key": alice_pubkey,
+    ///         "name": "Alice",
+    ///         "balance": "100",
+    ///     }, {
+    ///         "pub_key": bob_pubkey,
+    ///         "name": "Bob",
+    ///         "balance": "100",
+    ///     }])
+    /// );
+    /// # }
+    /// ```
+    @(ID = "wallets")
+    pub GetWallets(()) -> Vec<Wallet>;
```

Comment:
Can we infer `ID` from the name of endpoint itself?

Reply:
In my opinion syntax like `@ (ID =" wallets ")` is inconsistent and a quite unclear. I propose to use rust attribute syntax:
```rust
#[exonum_api(id = "wallets")]
pub get_wallets() -> Vec<Wallet>
```